### PR TITLE
ISPN-10558 ServerNG add database rules

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-10558

This now runs server DB tests against containerized H2, MySQL and PostgreSQL.

### CODE REVIEW

File: server/runtime/TESTING.md [hunk 1]
```diff
@@ -108,15 +108,49 @@ to pause a node and subsequently resume it.
...
+`mvn -DdefaultExcludedJUnitGroups=org.infinispan.server.test.category.Persistence ...`
 
 ## Testsuite Properties
 
 The following is a list of properties which affect the build:
 
 * `org.infinispan.test.server.baseImageName` the base image to use for the server. Defaults to `jboss/base-jdk:11`.
 * `org.infinispan.test.server.driver`  the driver to use, `EMBEDDED` or `CONTAINER`. Defaults to the `EMBEDDED` driver.
+* `org.infinispan.test.server.extension.libs` locates artifact defined by G:A:V, you can pass a list of libraries (comma separeted) to be copied to the server. Only needed for container mode.
+* `org.infinispan.test.server.jdbc.database` database name to be used during persistence tests.
```
Comment (gustavolira) [inline server/runtime/TESTING.md]: NIT: the right parameter name is `org.infinispan.test.server.jdbc.databases`

File: server/runtime/src/test/java/org/infinispan/server/test/persistence/DatabaseServerRule.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @author Gustavo Lira &lt;glira@redhat.com&gt;
+ * @since 10.0
+ **/
+public class DatabaseServerRule implements TestRule {
+   public static final String DATABASE = "org.infinispan.test.server.jdbc.database";
+   private final InfinispanServerRule infinispanServerRule;
+   private final String dbName;
+   public Database database;
+
+   public DatabaseServerRule(InfinispanServerRule infinispanServerRule, String dbName) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/test/persistence/DatabaseServerRule.java]: I've changed the way this is implemented by going through the DatabaseServerRule.getDatabaseTypes(...) method

File: server/runtime/TESTING.md [hunk 3]
```diff
@@ -116,7 +116,41 @@ The following is a list of properties which affect the build:
...
+###### DB2
+`mvn -pl server/runtime test -Dtest=PersistenceIT -Pcontainer -Dorg.infinispan.test.server.extension.libs=com.ibm.db2:jcc:11.1.4.4 -Dorg.infinispan.test.server.jdbc.database=db2`
+
+###### MARIADB
+`mvn -pl server/runtime test -Dtest=PersistenceIT -Pcontainer -Dorg.infinispan.test.server.extension.libs=org.mariadb.jdbc:mariadb-java-client:2.4.4 -Dorg.infinispan.test.server.jdbc.database=mariadb`
+
+###### SQL SERVER
+`mvn -pl server/runtime test -Dtest=PersistenceIT -Pcontainer -Dorg.infinispan.test.server.extension.libs=com.microsoft.sqlserver:mssql-jdbc:7.4.1.jre11 -Dorg.infinispan.test.server.jdbc.database=sql_server`
+
+###### ORACLE XE
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/TESTING.md]: I did update all of those files to be "EXTERNAL". While I haven't tested them on the real thing, they should be correct (and you can correct them in a later PR)

File: server/runtime/src/test/resources/configuration/datasource/mariadb.properties [hunk 4]
```diff
@@ -0,0 +1,3 @@
+id.column.type=VARCHAR(255)
+data.column.type=VARBINARY(1000)
+timestamp.column.type=BIGINT
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/resources/configuration/datasource/mariadb.properties]: Fixed.

File: server/runtime/src/main/java/org/infinispan/server/Extensions.java [hunk 5]
```diff
@@ -48,6 +48,12 @@ public void apply(HotRodServer server) {
       }
    }
 
+   private <T> void load(ClassLoader classLoader, Class<T> contract) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/Extensions.java]: Fixed

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/connectionfactory/PooledConnectionFactory.java [hunk 6]
```diff
@@ -51,6 +51,7 @@ public void start(ConnectionFactoryConfiguration config, ClassLoader classLoader
          if (propsFile != null) {
             dataSource = AgroalDataSource.from(new AgroalPropertiesReader(PROPERTIES_PREFIX).readProperties(propsFile));
          } else {
+            Class<?> driverClass = Util.loadClass(poolConfig.driverClass(), null);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/connectionfactory/PooledConnectionFactory.java]: Fixed

File: server/runtime/src/test/java/org/infinispan/server/persistence/PooledConnectionOperations.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * @since 10.0
+ **/
+@Category(Persistence.class)
+public class PooledConnectionOperations {
+
+   @ClassRule
+   public static InfinispanServerRule SERVERS = PersistenceIT.SERVERS;
+
+   @ClassRule
+   public static DatabaseServerRule DATABASE = new DatabaseServerRule(SERVERS, "postgres");
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/persistence/PooledConnectionOperations.java]: Fixed

File: server/runtime/TESTING.md (hunk 1 above)
Comment (tristantarrant) [reply server/runtime/TESTING.md]: Fixed

**Comment by wburns** [comment]
Failures in the new test class https://ci.infinispan.org/job/Infinispan/job/PR-7436/6/#showFailuresLink

**Comment by tristantarrant** [comment]
@wburns CI is now fixed (after Docker upgrade)

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/DatabaseType.java [hunk 8]
```diff
@@ -7,9 +7,9 @@
  * @since 4.1
  */
 public enum DatabaseType {
-   MARIA_DB, MYSQL, POSTGRES, DERBY, HSQL, H2, SQLITE,
+   MARIA_DB, MYSQL, MARIADB, POSTGRES, DERBY, HSQL, H2, SQLITE,
```
Comment (ryanemerson) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/DatabaseType.java]: Why do we have two enum values for MariaDB?

Comment (gustavolira) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/DatabaseType.java]: I added this new value but the logic that I created is no longer being used, this value can be removed.

File: server/runtime/pom.xml [hunk 9]
```diff
@@ -20,6 +20,7 @@
       <defaultExcludedTestGroup />
       <server.output.dir>${project.build.directory}/${infinispan.brand.prefix}-server-${infinispan.brand.version}</server.output.dir>
       <server.test.lib.dir>${project.build.directory}/testlibs</server.test.lib.dir>
+      <version.shrinkwrap.resolvers>3.1.3</version.shrinkwrap.resolvers>
```
Comment (ryanemerson) [inline server/runtime/pom.xml]: Should be defined in `build-configuration/pom.xml`

Comment (ryanemerson) [reply server/runtime/pom.xml]: Actually, we already have `${org.jboss.shrinkwrap.resolver}` defined in ^ pom and it's on version 2.2.4. Can't we upgrade that to 3.1.3 and reuse it?

We can then just add the following dep in this pom:

```xml
         <dependency>
            <groupId>org.jboss.shrinkwrap.resolver</groupId>
            <artifactId>shrinkwrap-resolver-depchain</artifactId>
            <type>pom</type>
            <scope>test</scope>
         </dependency>
```

Comment (gustavolira) [reply server/runtime/pom.xml]: This is definitely a better approach but I don't know the side effect of changing version for the whole project

Comment (ryanemerson) [reply server/runtime/pom.xml]: Let's find out https://github.com/infinispan/infinispan/pull/7458 :slightly_smiling_face:

**Comment by wburns** [comment]
Whoops thought I submitted these yesterday

File: server/runtime/TESTING.md [hunk 10]
```diff
@@ -108,15 +108,49 @@ to pause a node and subsequently resume it.
...
+
+`mvn -DdefaultExcludedJUnitGroups=org.infinispan.server.test.category.Persistence ...`
 
 ## Testsuite Properties
 
 The following is a list of properties which affect the build:
 
 * `org.infinispan.test.server.baseImageName` the base image to use for the server. Defaults to `jboss/base-jdk:11`.
 * `org.infinispan.test.server.driver`  the driver to use, `EMBEDDED` or `CONTAINER`. Defaults to the `EMBEDDED` driver.
+* `org.infinispan.test.server.extension.libs` locates artifact defined by G:A:V, you can pass a list of libraries (comma separeted) to be copied to the server. Only needed for container mode.
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/TESTING.md]: WILL ! your Java developer license is revoked !
It's a Maven coordinate: Group:Artifact:Version (e.g. com.h2database:h2:1.4.199)

File: server/runtime/src/test/java/org/infinispan/server/test/category/Persistence.java [hunk 11]
```diff
@@ -0,0 +1,8 @@
+package org.infinispan.server.test.category;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 10.0
+ **/
+public class Persistence {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/test/java/org/infinispan/server/test/category/Persistence.java]: This is used as a JUnit category (which usually are empty classes)

File: server/runtime/TESTING.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (wburns) [reply server/runtime/TESTING.md]: I am not a Maven wizard :)

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/DatabaseType.java (hunk 8 above)
Comment (tristantarrant) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/DatabaseType.java]: I have reformatted this class to remove the duplicate, put each db on a single line and sort them

**Comment by tristantarrant** [comment]
Included the shrinkwrap upgrade as well as the MariaDB fix

File: server/runtime/pom.xml (hunk 9 above)
Comment (tristantarrant) [reply server/runtime/pom.xml]: I have folder that PR in here and adjusted other dependencies in the process.

**Comment by ryanemerson** [comment]
Thanks @gustavolira @tristantarrant

**Pull request merged**